types/field_type.go

```diff
@@ -250,54 +250,70 @@ func (ft *FieldType) Restore(ctx *format.RestoreCtx) error {
 	return nil
 }
 
-// FormatAsCastType is used for write AST back to string.
-func (ft *FieldType) FormatAsCastType(w io.Writer) {
+// RestoreAsCastType is used for write AST back to string.
+func (ft *FieldType) RestoreAsCastType(ctx *format.RestoreCtx) {
 	switch ft.Tp {
 	case mysql.TypeVarString:
 		if ft.Charset == charset.CharsetBin && ft.Collate == charset.CollationBin {
-			fmt.Fprint(w, "BINARY")
+			ctx.WriteKeyWord("BINARY")
 		} else {
-			fmt.Fprint(w, "CHAR")
+			ctx.WriteKeyWord("CHAR")
 		}
 		if ft.Flen != UnspecifiedLength {
-			fmt.Fprintf(w, "(%d)", ft.Flen)
+			ctx.WritePlainf("(%d)", ft.Flen)
 		}
 		if ft.Flag&mysql.BinaryFlag != 0 {
-			fmt.Fprint(w, " BINARY")
+			ctx.WriteKeyWord(" BINARY")
 		}
-		if ft.Charset != charset.CharsetBin && ft.Charset != mysql.DefaultCharset {
-			fmt.Fprintf(w, " CHARACTER SET %s", ft.Charset)
+		if ctx.Flags.HasForGeneratedColumnFlag() {
+			if ft.Charset != charset.CharsetBin {
+				ctx.WriteKeyWord(" CHARSET ")
+				ctx.WritePlain(strings.ToLower(ft.Charset))
```

Comment:
We usually use `WriteKeyWord` directly for charset and collations.

```suggestion
				ctx.WriteKeyWord(ft.Charset)
```

Reply:
@kennytm PTAL. Thanks.
I wondered parsing `CharsetName` as lower is a good idea or not.

Another solution is changing some UTs. e.g:
functions_test.go, TestFuncCastExprRestore
```sql
{"CONVERT('Müller', CHAR(32) CHARACTER SET UtF8)", "CONVERT('Müller', CHAR(32) CHARACTER SET UTF8)"}
```
----->
```sql
{"CONVERT('Müller', CHAR(32) CHARACTER SET UTF8)", "CONVERT('Müller', CHAR(32) CHARACTER SET UTF8)"}
```

Reply:
@pingyu We could create a dedicated `WriteCharSet` function which writes the input as lowercase, but this is mostly cosmetic change that doesn't affect the validity of the actual output. We could enhance that in future PRs.